### ISSUE TITLE
Make http requests from the service

### ISSUE THREAD

**Issue by areina** [mauron85/cordova-plugin-background-geolocation#20]
Hi!
If I understand correctly how the plugin works, when the app is in background (or killed by the user), the service stores user locations in a database and when the app is opened again it's possible to send these locations to a remote server.

This is fine for some use cases, but do you think it would be possible to make http requests sending user location from the service?

Thank you

**Comment by mauron85** [comment]
It only stores when app (main activity) is killed by the system. If your app is in the background it keep receiving location updates.

I actually thrown out the android implementation for making http posts directly from the plugin.

Do you really need it? My point is that there is only one situation that it makes sense and it's the killed by the system. But from my experience on recent devices (with lot of ram), my test app was never killed by the system and to reproduce such a conditions I had to kill app from terminal.

**Comment by areina** [comment]
I'm developing an application that needs to send user location continuously, and it should keep doing it when the app is closed (swiped from the applications list). As far I know, in that case, it's possible to maintain the service running but the main activity doesn't exist, so it should be the service who makes the http request.
Probably this shouldn't be the default behaviour, but I think it would be nice if it's a configurable option.
By the way, this would be possible only if #9 is solved.

**Comment by Kujbor** [comment]
I've made for myself storing data on a server using HttpURLConnection in AsyncTask, but it would be great to see it in your plugin, so... +1!)

**Comment by mauron85** [comment]
@Kujbor go ahead push your PR. ;-)

**Comment by Kujbor** [comment]
I did it for my realization. But I'll check your implementation, and if it works, then transferred to your plugin and push commit)

**Comment by mauron85** [comment]
@areina @Kujbor Any suggestions about configuration and message POST format?
I would like to make it as simple as possible.

Example:
```
backgroundGeoLocation.configure(callbackFn, failureFn, {
  postUrl: 'http://some.url/location',
  postAlways: true // false = post locations only when main activity is dead
});
```

And format of posted message would be same json as received by callbackFn.

``` json
{
  "locationId": 1,
  "time": 1450738145721,
  "latitude": 49.1633779,
  "longitude": 20.2710556,
  "accuracy": 40.5,
  "speed": 0,
  "altitude": 0,
  "bearing": 0,
  "serviceProvider": "ANDROID_DISTANCE_FILTER",
  "debug": true
}
```

**Comment by areina** [comment]
Maybe for first iteration is fine to keep it simple, but I'm thinking that some kind of authentication for those requests will be needed and then we should be able to configure headers or params. 
About message POST format, I think it's ok.

Thank you for your work!

**Comment by Kujbor** [comment]
What may come in handy this functionality without authorization?)) In my implementation, I have sent a request with cookies, that I have taken from webview. In addition, I use get instead of post, because the server is using csrf-protection, and it is much easier to using get-request than an implementation of mechanism for obtaining ssrf token.

**Comment by Kujbor** [comment]
mauron85, when I asked in the problem #9 where I can place in your application code to save data on the server, I meant exactly what I asked) I have implemented (in Java for Android) storing data on the server using cookies, I want to understand where I can insert it in your plugin so that it is guaranteed to be invoked at a location update in the background, and in the foreground (for the Android).

**Comment by christocracy** [comment]
If you want to save yourself time/money, I've figured out these problems and many more (eg: Android start-on-boot, integrated geofencing).  I'm the original author of `cordova-background-geolocation`, from which this project is a fork of and continues to borrow its ideas from.

http://shop.transistorsoft.com/pages/cordova-background-geolocation-premium

Why is integrated geofencing important?  Why not do what [this plugin recommends](https://github.com/mauron85/cordova-plugin-background-geolocation#geofencing) and use an existing 3rd party geofencing plugin?  For the case when your plugin is running in *headless* mode, where the user has closed the app but you still want to continue tracking (ie: `stopOnTerminate: true` -- for this case, the geofencing implementation **must** be integrated within the plugin.

**Comment by Kujbor** [comment]
christocracy, I saw your implementation of this functionality, and perhaps that is exactly what I need. But unfortunately I am from Russia. The country is in crisis, and your implementation is prohibitively expensive for me. I trite not be able to spend a quarter of my monthly salary for the library for a non-profit app for personal use. So I quite like everything, but I can not buy. Thank you)

**Comment by clembo590** [comment]
I would like to say "+1" for this feature: sending the position to the server while in background or even when the app is killed by the OS is a must.
BUT this feature is useless if we cannot configure the data we are sending from the service.
can you add an  "optional" param that we could fill with whatever we want to our server.
backgroundGeoLocation.configure(callbackFn, failureFn, {
  postUrl: 'http://some.url/location',
  postAlways: true, // false = post locations only when main activity is dead
  optional: anotherJson
});

or do not re invent the wheel... just use the data structure christocray used in his plugin:

        // HTTP / SQLite config
        url: 'http://posttestserver.com/post.php?dir=cordova-background-geolocation',
        method: 'POST',
        batchSync: true,       // <-- [Default: false] Set true to sync locations to server in a single HTTP request.
        autoSync: true,         // <-- [Default: true] Set true to sync each location to server as it arrives.
        maxDaysToPersist: 1,    // <-- Maximum days to persist a location in plugin's SQLite database when HTTP fails
        headers: {
            "X-FOO": "bar"
        },
        params: {
            "auth_token": "maybe_your_server_authenticates_via_token_YES?"
        }

**Comment by christocracy** [comment]
This plugin hasn't been supported for over a year now.

It has moved [Here](https://github.com/transistorsoft/cordova-background-geolocation-lt).

This feature you require, sending HTTP requests from the service is available there.

**Comment by christocracy** [comment]
(If anyone received my previous message (**deleted**), ignore it.  I thought I was commenting on my own repo)

**Comment by mauron85** [comment]
@clembo590 I would agree that headers are needed, for example if you're using some kind of authentication on the server (Bearer token header...). But IMHO params are unnecessary. You can do the same thing by creating url with params such as http://yourserver.tld/enpoint?param1=val1&param2=arg2. You can adjust server to read those params easily. It would make sense if params could be set individual per request, but that is not the case.

**Comment by ElieSauveterre** [comment]
I have done something similar on the Geofence Plugin https://github.com/ElieSauveterre/cordova-plugin-geofence/blob/72848cb3e3bde3af8e0e6a2a5c4c5dbbf9563c6c/src/android/TransitionReceiver.java

I think that the plugin should send location updates when the app is running and when in background et terminated. Headers would be great.

**Comment by uzumakinaruto123** [comment]
@areina  any success in making http requests when app not running? i have the same requirement where application needs to send user location continuously.

Thanks ! help is appreciated

**Comment by leobudima** [comment]
+1, this would be extremely useful for my use-case

**Comment by christocracy** [comment]
I've been doing HTTP in both iOS & Android for over a year in the original [Background Geolocation](https://github.com/transistorsoft/cordova-background-geolocation-lt/tree/master/docs#http--persistence-options), of which this project is a fork.

**Comment by leobudima** [comment]
@christocracy - Hi Chris - does your plugin do it from the service, even when app is killed by the system (on both platforms)? Thanks

**Comment by christocracy** [comment]
@leobudima yes.

**Comment by mauron85** [comment]
Actually guys, why not support christocracy by buing his product. This plugin would probably never exists without his hard work.

**Comment by christocracy** [comment]
Thanks @mauron85.  Here's a **25 % off** coupon for first five.  Purchase [here](http://shop.transistorsoft.com/pages/cordova-background-geolocation-premium#plans)
```
mauron85-issue-20
```

**Comment by leobudima** [comment]
Of course - I think it's great work, obviously a lot of effort has gone into it, has great features, great support and it's well worth buying. As a fellow developer, I obviously wish success to him and good sales!

However, in the economy I live/work in, one sometimes needs to find alternatives - and that's why I hugely appreciate your effort, @mauron85, and want to support it where I can. I'm looking into getting on top of some of the problems we have with tracking in our app, and then I'll try to see if I can help with some of the issues or by submitting PRs for additional features we need and use, that others might benefit from.

**Comment by mauron85** [comment]
Sure. I love PRs. Don't hesitate. :-)

**Comment by netbitsolutions** [comment]
@mauron85 , we appreciate your hard work so far and will hang on till we can have the posting to server when app is on background. When will this be possible ?

Also can you share a sample code on how to extract locations saved in  when app is on background, you once shared this code
`document.addEventListener('deviceready', onDeviceReady, false);
function onDeviceReady () {
    backgroundGeoLocation.getLocations(function(locations) {
        [].filter.call(locations, function(location) {
            if (location.debug === false) {
                 //do ajax post of location
                 backgroundGeoLocation.deleteLocation(location.locationId);
            };
        });
    });
}
` 

Am yet not able to fetch the saved  locations pls help

**Comment by mauron85** [comment]
Hi guys. Next branch currently has very basic support for http posting for Android. Read more in my comment.

https://github.com/mauron85/cordova-plugin-background-geolocation/issues/59#issuecomment-223319855

**Comment by mauron85** [comment]
Anybody willing to try autosync for iOS?

Install next version 3.0.0-alpha.x
```
cordova plugin add cordova-plugin-mauron85-background-geolocation@next
```
DOCS:
https://github.com/mauron85/cordova-plugin-background-geolocation/tree/next-ios

**Comment by mauron85** [comment]
Implemented in version 2.2.0
